core/unit_test/tools/TestKernelNames.cpp

```diff
@@ -0,0 +1,170 @@
+//
+//@HEADER
+
+#include <Kokkos_Core.hpp>
+
+#include <gtest/gtest.h>
+
+namespace {
+
+std::string last_parallel_for;
+std::string last_parallel_reduce;
+std::string last_parallel_scan;
+
+void get_parallel_for_kernel_name(char const* kernelName, uint32_t /*deviceID*/,
+                                  uint64_t* /*kernelID*/) {
+  last_parallel_for = kernelName;
+}
+
+void get_parallel_reduce_kernel_name(char const* kernelName,
+                                     uint32_t /*deviceID*/,
+                                     uint64_t* /*kernelID*/) {
+  last_parallel_reduce = kernelName;
+}
+
+void get_parallel_scan_kernel_name(char const* kernelName,
+                                   uint32_t /*deviceID*/,
+                                   uint64_t* /*kernelID*/) {
+  last_parallel_scan = kernelName;
+}
+
+struct WorkTag {};
+
+void test_kernel_name_parallel_for() {
+  Kokkos::Tools::Experimental::set_begin_parallel_for_callback(
+      get_parallel_for_kernel_name);
+
+  using ExecutionSpace = Kokkos::DefaultExecutionSpace;
+
+  {
+    std::string my_label = "my_parallel_for_range_policy";
+
+    auto my_lambda = KOKKOS_LAMBDA(int){};
+    Kokkos::parallel_for(my_label, Kokkos::RangePolicy<ExecutionSpace>(0, 1),
+                         my_lambda);
+    ASSERT_EQ(last_parallel_for, my_label);
+
+    Kokkos::parallel_for(Kokkos::RangePolicy<ExecutionSpace>(0, 1), my_lambda);
+    ASSERT_EQ(last_parallel_for, typeid(my_lambda).name());
+
+    auto my_lambda_with_tag = KOKKOS_LAMBDA(WorkTag, int){};
+    Kokkos::parallel_for(my_label,
+                         Kokkos::RangePolicy<ExecutionSpace, WorkTag>(0, 1),
+                         my_lambda_with_tag);
+    ASSERT_EQ(last_parallel_for, my_label);
+
+    Kokkos::parallel_for(Kokkos::RangePolicy<ExecutionSpace, WorkTag>(0, 1),
+                         my_lambda_with_tag);
+    ASSERT_EQ(last_parallel_for,
+              std::string(typeid(my_lambda_with_tag).name()) + "/" +
+                  typeid(WorkTag).name());
+  }
+
+  Kokkos::Tools::Experimental::set_begin_parallel_for_callback(nullptr);
+}
+
+void test_kernel_name_parallel_reduce() {
+  Kokkos::Tools::Experimental::set_begin_parallel_reduce_callback(
+      get_parallel_reduce_kernel_name);
+
+  using ExecutionSpace = Kokkos::DefaultExecutionSpace;
+
+  {
+    std::string my_label = "my_parallel_reduce_range_policy";
+    float my_result;
+
+    auto my_lambda = KOKKOS_LAMBDA(int, float&){};
+    Kokkos::parallel_reduce(my_label, Kokkos::RangePolicy<ExecutionSpace>(0, 1),
+                            my_lambda, my_result);
+    ASSERT_EQ(last_parallel_reduce, my_label);
+
+    Kokkos::parallel_reduce(Kokkos::RangePolicy<ExecutionSpace>(0, 1),
+                            my_lambda, my_result);
+    ASSERT_NE(last_parallel_reduce.find(typeid(my_lambda).name()),
+              std::string::npos)
+        << last_parallel_reduce << " does not contain "
+        << typeid(my_lambda)
+               .name();  // internally using
```

Comment:
Wouldn't the test output in case of failure better if using `HasSubstr` from https://google.github.io/googletest/reference/matchers.html#string-matchers ?

Reply:
Along these lines:
```suggestion
    ASSERT_THAT(last_parallel_reduce, ::testing::HasSubstr(typeid(my_lambda).name()));
```

Reply:
That is a good suggestion, except we can't use GoogleMock yet

Reply:
I am generally in favor of adding GMock but this likely would not play well with the ongoing TriBITS refactor and would be getting somewhat trivial after we are done with it